test/e2e/plain-v0_provisioner_test.go

```diff
@@ -345,6 +345,139 @@ var _ = Describe("plain-v0 provisioner bundleinstance", func() {
+		})
+		AfterEach(func() {
+			By("deleting the testing Bundle resource")
+			Eventually(func() error {
+				return client.IgnoreNotFound(c.Delete(ctx, bundle))
+			}).Should(Succeed())
+
+			By("deleting the testing BundleInstance resource")
+			Eventually(func() error {
+				return client.IgnoreNotFound(c.Delete(ctx, bi))
+			}).Should(Succeed())
+		})
+
+		It("should project a failed installation state", func() {
+			By("projecting a failed installation state")
+			Eventually(func() bool {
+				if err := c.Get(ctx, client.ObjectKeyFromObject(bi), bi); err != nil {
+					return false
+				}
+				if bi.Status.InstalledBundleName != "" {
+					return false
+				}
+				existing := meta.FindStatusCondition(bi.Status.Conditions, "Installed")
+				if existing == nil {
+					return false
+				}
+				// TODO(tflannag): Add a custom error type for API-based Bundle installations that
+				// are missing the requisite CRDs to be able to deploy the unpacked Bundle successfully.
+				// TODO(tflannag): This kind of comparison feels brittle. Also: there may be cache
+				// invalidation issues for the GVKs we watch, or there's something non-deterministic
+				// happening under-the-hood when lifecycling through helm's implementation. To reproduce,
+				// run this check against an existing cluster, focus on this test, and continue testing
+				// until it results in a different message.
+				expectedMessage := `
+				error while running post render on files: [unable to recognize "": no
+				matches for kind "CatalogSource" in version "operators.coreos.com/v1alpha1",
+				unable to recognize "": no matches for kind "ClusterServiceVersion" in version
+				"operators.coreos.com/v1alpha1", unable to recognize "": no matches for kind
+				"OLMConfig" in version "operators.coreos.com/v1", unable to recognize "": no
+				matches for kind "OperatorGroup" in version "operators.coreos.com/v1"]
+				`
+				expected := metav1.Condition{
+					Type:    "Installed",
+					Status:  metav1.ConditionStatus(corev1.ConditionFalse),
+					Reason:  "InstallFailed",
+					Message: strings.Join(strings.Fields(expectedMessage), " "),
+				}
+				return conditionsSemanticallyEqual(expected, *existing)
+			}).Should(BeTrue())
+
+			By("creating the dependent OLM CRDs Bundle resource")
+			bundleCRDs := &rukpakv1alpha1.Bundle{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "olm-crds",
+				},
+				Spec: rukpakv1alpha1.BundleSpec{
+					ProvisionerClassName: plainProvisionerID,
+					Image:                "quay.io/tflannag/olm-plain-bundle:olm-crds-v0.20.0",
+				},
+			}
+			err := c.Create(ctx, bundleCRDs)
+			Expect(err).To(BeNil())
+
+			biCRDS := &rukpakv1alpha1.BundleInstance{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "olm-crds",
+				},
+				Spec: rukpakv1alpha1.BundleInstanceSpec{
+					ProvisionerClassName: plainProvisionerID,
+					BundleName:           bundleCRDs.GetName(),
+				},
+			}
+			err = c.Create(ctx, biCRDS)
+			Expect(err).To(BeNil())
```

Comment:
Note: I was having a hard time structuring this e2e spec and wanted to throw this up now to get some opinions. Right now, the CRD Bundle/BI testing resources get created in a test spec (vs. in a BeforeEach/AfterEach setup) and using the defer keyword in ginkgo seems like a poor idea given potential consistency issues.

Reply:
The pattern I've generally seen is that `BeforeEach` and `AfterEach` are purely for setup/teardown for the parent's `Describe`'d scenario.

The the sibling `It`s are the various things you want to test in that scenario. There also the possibility of nesting more Describe/Before|AfterEach/It scopes on child layers to build a sort of test matrix.

For this particular `Describe`, I might have done something like this:
```go 
Describe("when a BI has a dependency", func() {
	BeforeEach() // setup the BI that has a dependency
	AfterEach() // teardown the BI that has a dependency
	When("the dependency is unmet", func() {
		It("should propagate a failure to the BI")
	})
	When("the dependency is met", func() {
		BeforeEach() // setup the dependency
		AfterEach() // teardown the dependency
		It("should successfully install the BI")
	})
	When("the dependency is initially unmet", func() {
		It("should eventually succeed installing the BI after the dependent is installed", func() {
			By("waiting for the failure")
			By("installing the dependency")
			By("waiting for success")
		})
	})
})
```

These may seem repetitive and overkill, but I've found:
- its easier to add new tests to existing scenarios later
- you can catch hard to find bugs by forcing the controller to permute more state transitions
    - nil to failure
    - nil to direct success
    - nil to failure to success

It's not super obvious now, but when we add more conditions and other status fields, I think the value will become more clear.